### ISSUE TITLE
custom attribute in filename isn't working

### ISSUE THREAD

**Issue by Kaiz3rTool** [boberfly/GafferCycles#20]
If i put a custom attribute in a filename field from "image texture" node, it doesn't work and render a pink shader
the exact same expression works as expected  with appleseed

not sure it's related but starting gaffer I have this error message : 

> which: no nvcc in (/opt/gaffer-0.54.2.0-linux/appleseed/bin:/opt/gaffer-0.54.2.0-linux/bin:/opt/gaffer-0.54.2.0-linux/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/william/.local/bin:/home/william/bin:/opt/Arnold/bin:/opt/3delight-1.4.1/bin:/mnt/Pipeline/Gaffer/deadline/bin:/opt/gafferCycles/bin)

it seems to be a NVIDIA problem, this message appear only if gaffercycles is loaded at startup

best

**Comment by boberfly** [comment]
Hi @Kaiz3rTool 
I'd recommend checking what Cycles supports for textures, it probably doesn't have the same custom attribute support (do you mean <udim>? I need more info to go by). 
Cheers

**Comment by Kaiz3rTool** [comment]
Hi, the custom attrib is only for the path of the texture, there is no udim for now
I test the process with jpeg file so it should be ok

here is an example scene with some notes to illustrate the problem

bests

[cyclesCustomAttrib.gfr.txt](https://github.com/boberfly/GafferCycles/files/3742431/cyclesCustomAttrib.gfr.txt)

**Comment by boberfly** [comment]
Oh I see what you mean, that's a very new feature actually for Gaffer, in fact I think only 0.54.2.0 has it! https://github.com/GafferHQ/gaffer/releases

It could be because I am still building against 0.54.1.0 here and I need to update on my end, I'll look into making a build tonight with 0.54.2.0 with the OSL fix for https://github.com/boberfly/GafferCycles/issues/21

**Comment by boberfly** [comment]
I quickly saw that the Renderer backends for Arnold/Appleseed/etc. got an update to support shader substitutions, so I added the same thing but haven't tested yet:
https://github.com/boberfly/GafferCycles/releases/tag/0.10.0

**Comment by Kaiz3rTool** [comment]
thanks, I will test it soon and let you know

**Comment by boberfly** [comment]
I did notice you were probably struggling to get a HDRI to function in Cycles. I want to clean this up, but you can achieve it by placing a CyclesBackground node, and connecting a background_shader to this and an environment texture. The mapping on the environment texture needs to be X Z Y (Z & Y flipped) and scale to x -1.0, y 1.0, z 1.0 to have the same mapping as Appleseed. An actual background_light placed is merely used to grab the importance sampling settings as of right now and not used otherwise.

Long story short, Blender only ever supports one HDRI at a time, but Cycles lets you create multiple background_light types, but the codebase I don't think has ever been used to do this and needs further investigating and patches to allow this.

Cheers!

**Comment by Kaiz3rTool** [comment]
I still have the same issue for now with the attribute

Sorry i didn't understand how to reproduce the hdri workflow, as i cant connect the cyclesbackground and the background shader. do you have a small graph to illustrate it please ? 
thanks

cheers

**Comment by boberfly** [comment]
Uploaded a hdri.gfr to show what I mean (yeah not too elegant for now).
[hdri.gfr.txt](https://github.com/boberfly/GafferCycles/files/3747331/hdri.gfr.txt)

**Comment by Kaiz3rTool** [comment]
Hi, thanks for the help, I get it now
I have made some test, I'm having better results when I remove the " -2* " from the rotation expression. without this multiplication, it seems that rotation on background_light match what we expect to see

`import math
import imath

rot_x = - math.pi * ( parent["background_light"]["transform"]["rotate"]["x"] / 360.0 )
rot_y = math.pi * ( parent["background_light"]["transform"]["rotate"]["z"] / 360.0 ) 
rot_z = math.pi * ( parent["background_light"]["transform"]["rotate"]["y"] / 360.0 )

parent["environment_texture2"]["parameters"]["tex_mapping__rotation"] = imath.V3f( rot_x, rot_y, rot_z )`

the minus in rot_x is because i think that if we invert x in scale, the rotation need to be inverted too (i need to double check that)


about the original thread: attribute in filename, it's work as expected, my test image wasn't there for any reason so my previous test wasn't right, my bad



thanks again for your help !

**Comment by boberfly** [comment]
Awesome! As long as it matches the other renderers and visualiser and works for you all good, the -1 is because the image should be inverted as the sphere is the one capturing the light and not the light visual itself.

I have some more fixes coming, as I noticed the convert_X_to_Y shaders weren't getting translated right when I tried them.
